whylog/teacher/__init__.py

```diff
@@ -0,0 +1,70 @@
+class Interval(object):
+    """
+    Represents interval in line.
+    """
+
+    def __init__(self, start_offset, length, abstract_line):
+        self.start_offset = start_offset
+        self.length = length
+        self.abstract_line = abstract_line
+
+
+class Rule(object):
+    """
+    Represents rule entering by user.
+    """
+
+    def __init__(self, effect):
+        self.effect = effect
+        self.causes = set()
+
+
+class Teacher(object):
+    """
+    Enable teaching new rule. One Teacher per one entering rule.
+    """
+
+    def __init__(self, effect, causes=None):
+        self._rule = Rule(effect)
+        if causes is not None:
```

Comment:
```py
self._rule.causes.update(causes)
```
is an operation on `self._rule.causes` set adds element from `causes` to the `self._rule.causes` set. If default `causes` is an empty frozendict or an empty tuple, `self._rule.causes.update(causes)` is a noop and that is why the if statement is not necessary.

But causes could not be an empty list or a set. Do you know why?